### PR TITLE
Support multiple rights vs same left scenario with a simple call to ApiComparer

### LINKED ISSUE #16918: ApiCompatibility.RuleDriver logic to load the rules and run them on the APIs should be revisited
PR https://github.com/dotnet/sdk/pull/16817 added the ApiCompatibility object model and currently the way the rules are executed is for every `ElementMapper` visited in the `DifferenceVisitor` we iterate every rule that we resolved via reflection and invoke `Rule.Run(ElementMapper)`. 

This is the "hot" path of the object model and we should run numbers and see if there is a better way to do this, there might be other ways like events/hooks that might serve better.

https://github.com/dotnet/sdk/pull/16817#discussion_r612602004 for more context.

cc: @ericstj @Anipik

### PR DESCRIPTION
**Pull request by safern**
With these changes, comparing the same left against multiple rights is optimized ~30% in time, here are the benchmark results:

|                 Method |     Mean |   Error |  StdDev |
|----------------------- |---------:|--------:|--------:|
| UseOneApiCompatPerPath | 442.8 ms | 5.12 ms | 4.79 ms |
|      CompareOnePerSide | 107.5 ms | 0.64 ms | 0.53 ms |
| UseOneApiCompatOverall | 291.4 ms | 2.05 ms | 1.92 ms |

`UseOneApiCompatPerPath` is invoking `ApiComparer.GetDifferences` for each combination of (left, right) which takes `442.8ms`, vs now with this changes you would only need to do what I do in `UseOneApiCompatOverall` which calls `ApiComparer.GetDifferences` once with an `IList` for the rights, which takes `291.4ms` for comparing `System.Runtime.dll` ref assembly vs `System.Runtime.dll` implementation with 4 different instances of it, that's a `151ms` improvement.

Also, the regular scenario of comparing one left against one right, is about the same, this is before these changes:
|                 Method |     Mean |   Error |  StdDev |
|----------------------- |---------:|--------:|--------:|
|      CompareOnePerSide | 101.7 ms | 1.03 ms | 0.86 ms |

Here are the benchmarks I used to measure this: https://gist.github.com/safern/6a1354ee3e3d1e13272c5c3119a6e466

TODO after this change is being able to customize the `NoWarn` and `IgnoredDifferences` for the different rights that are being passed as at the moment it uses the same, which might not be desirable in some cases. I'll address that as part of the error baselining story.

Also, change how the Rules are loaded and invoked, by following a similar model as Roslyn, where an Initialize method is invoked and we pass in a context. Then Rules can register to context callbacks, that way the Rule just register to whatever callbacks it is important to run on based on the Mappers it runs for. This change fixes: https://github.com/dotnet/sdk/issues/16918

### CODE REVIEW

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/ApiComparer.cs [hunk 1]
```diff
@@ -47,6 +26,8 @@ public ApiComparer(bool includeInternalSymbols)
         /// </summary>
         public (string diagnosticId, string memberId)[] IgnoredDifferences { get; set; }
 
+        public Func<ComparingSettings> GetComparingSettings { get; set; }
```
Comment (safern) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/ApiComparer.cs:29]: I need to add docs for these changes. Will do that in a separate commit.

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/ElementContainer.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.ApiCompatibility.Abstractions
+{
+    public class ElementContainer<T>
```
Comment (safern) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/ElementContainer.cs:6]: I started with this as a Generic type even though it is now currently used with `IAssemblySymbol` only as this might be useful to use it to store the elements in the mappers and that way we could customize the diagnostic messages using the metadata information for each element in the mapper.

However I don't know if it is too much overhead and instead just have a setting to customize the `left` and `right` names when raising diagnostics.

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        public readonly string AssemblyName;
+        public readonly string TargetFramework;
+        public readonly string AssemblyType;
+
+        public MetadataInformation(string name, string targetFramework, string assemblyType)
+        {
+            AssemblyName = name ?? string.Empty;
+            TargetFramework = targetFramework ?? string.Empty;
+            AssemblyType = assemblyType ?? string.Empty;
```
Comment (Anipik) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs:22]: should we make this an enum ?

Comment (safern) [reply src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs:22]: How would that work? I was expecting this to have, `lib`, or `runtimes/win-x64`, etc. We would have to list all the possible runtimes?

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/ElementContainer.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        public ElementContainer(T element, MetadataInformation metadataInformation)
+        {
+            Element = element;
+            MetadataInformation = metadataInformation;
+        }
+
+        /// <summary>
+        /// The element that the container is holding.
+        /// </summary>
+        public T Element { get; set; }
```
Comment (Anipik) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/ElementContainer.cs:26]: should we make the setter private ?

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/Mappers/ElementSide.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.ApiCompatibility.Abstractions
+{
+    /// <summary>
+    /// Enum representing the side of an element in the mappers.
+    /// </summary>
+    public enum ElementSide : byte
```
Comment (Anipik) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/Mappers/ElementSide.cs:9]: curious why we are inheriting the byte class ?

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs (hunk 3 above)
Comment (Anipik) [reply src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs:22]: ah okay, i thought we are just storing the assembly type like, ref, lib or runtime

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/TypeMustExistTests.cs [hunk 6]
```diff
@@ -276,5 +279,271 @@ internal class InternalNested { }
...
+                    new CompatDifference(DiagnosticIds.TypeMustExist, $"Type 'CompatTests.First.FirstNested' exists on the left but not on the right", DifferenceType.Removed, "T:CompatTests.First.FirstNested"),
+                }),
+                (left.MetadataInformation, new MetadataInformation(string.Empty, string.Empty, "runtime-2"), new CompatDifference[]
+                {
+                    new CompatDifference(DiagnosticIds.TypeMustExist, $"Type 'CompatTests.First.FirstNested.SecondNested' exists on the left but not on the right", DifferenceType.Removed, "T:CompatTests.First.FirstNested.SecondNested"),
+                }),
+                (left.MetadataInformation, new MetadataInformation(string.Empty, string.Empty, "runtime-3"), new CompatDifference[]
+                {
+                    new CompatDifference(DiagnosticIds.TypeMustExist, $"Type 'CompatTests.First' exists on the left but not on the right", DifferenceType.Removed, "T:CompatTests.First"),
+                }),
```
Comment (Anipik) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/TypeMustExistTests.cs:429]: I think we should develop better design model so that we can add new tests relatively simpler. Not a blocker for pr or for preview, just something to think about

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs [hunk 7]
```diff
@@ -332,6 +333,164 @@ public class First
             };
 
             Assert.Equal(expected, differences);
+        }
+
+                [Fact]
```
Comment (Anipik) [inline src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/MembersMustExistTests.cs:338]: nit: indentation

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/RuleRunnerContext.cs [hunk 8]
```diff
@@ -0,0 +1,73 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
```
Comment (Anipik) [inline src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/RuleRunnerContext.cs:3]: nit : extra ```//```

**Review by Anipik — APPROVED** [review]

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs (hunk 3 above)
Comment (safern) [reply src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/MetadataInformation.cs:22]: No, I think it should be more specific to the runtime the assembly belongs to if it is a runtime assembly.

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/ElementContainer.cs (hunk 4 above)
Comment (safern) [reply src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/ElementContainer.cs:26]: Sure.

File: src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/Mappers/ElementSide.cs (hunk 5 above)
Comment (safern) [reply src/Compatibility/Microsoft.DotNet.ApiCompatibility/Abstractions/Mappers/ElementSide.cs:9]: Because `ElementSide` will never contain more than 2 values, so no need to use the default `int` underlying type for the enum fields, we can just use byte to use smaller fields memory wise.

File: src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/TypeMustExistTests.cs (hunk 6 above)
Comment (safern) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/TypeMustExistTests.cs:429]: I think the way I'm doing this expected vs actual can definitely be simplified. I'll look into simplifying it for the single left vs multiple left scenarios.

Comment (safern) [reply src/Tests/Microsoft.DotNet.ApiCompatibility.Tests/Rules/TypeMustExistTests.cs:429]: Discussed offline.
